src/sstruct_mv/sstruct_matrix.c

```diff
@@ -2173,20 +2135,254 @@ hypre_SStructMatrixToUMatrix( HYPRE_SStructMatrix  matrix,
       hypre_TFree(rows, HYPRE_MEMORY_DEVICE);
       hypre_TFree(rowidx, HYPRE_MEMORY_DEVICE);
       hypre_TFree(cols, HYPRE_MEMORY_DEVICE);
+      hypre_TFree(values, HYPRE_MEMORY_HOST);
    }
 
-   /* Set entries of ij_Ahat */
+   hypre_BoxArray ***convert_boxa;
+   convert_boxa = hypre_TAlloc(hypre_BoxArray **, nparts, HYPRE_MEMORY_HOST);
    for (part = 0; part < nparts; part++)
    {
-      pmatrix = hypre_SStructMatrixPMatrix(matrix, part);
+      pgrid = hypre_SStructGridPGrid(grid, part);
+      nvars = hypre_SStructPGridNVars(pgrid);
+      convert_boxa[part] = hypre_CTAlloc(hypre_BoxArray *, nvars, HYPRE_MEMORY_HOST);
+      for (var = 0; var < nvars; var++)
+      {
+         /* WM: question - what about connections between variables? Note the var, var arguments below */
+         sgrid      = hypre_SStructPGridSGrid(pgrid, var);
+         grid_boxes = hypre_StructGridBoxes(sgrid);
+         convert_boxa[part][var] = hypre_BoxArrayCreate(0, ndim);
+         hypre_ForBoxI(i, grid_boxes)
+         {
+            grid_box = hypre_BoxArrayBox(grid_boxes, i);
+            hypre_AppendBox(grid_box, convert_boxa[part][var]);
+         }
+      }
+   }
+   hypre_SStructMatrixBoxesToUMatrix(matrix, grid, &ij_Ahat, convert_boxa);
+
+   /* Free memory */
+   for (part = 0; part < nparts; part++)
+   {
+      pgrid = hypre_SStructGridPGrid(grid, part);
+      nvars = hypre_SStructPGridNVars(pgrid);
+      for (var = 0; var < nvars; var++)
+      {
+         hypre_BoxArrayDestroy(convert_boxa[part][var]);
+      }
+      hypre_TFree(convert_boxa[part], HYPRE_MEMORY_HOST);
+   }
+   hypre_TFree(convert_boxa, HYPRE_MEMORY_HOST);
+
+   HYPRE_ANNOTATE_FUNC_END;
+
+   return ij_Ahat;
+}
+
+/*--------------------------------------------------------------------------
+ * hypre_SStructMatrixUBoxes WM: todo - is this actually what I want?
+ *
+ * Get a set of boxes that covers points with non-zero rows in the
+ * unstructured matrix.
+ *
+ *--------------------------------------------------------------------------*/
+HYPRE_Int
+hypre_SStructMatrixUBoxes( hypre_SStructMatrix   *A,
```

Comment:
Should we move this to other branch and continue its development from there?

Reply:
Agreed. Removing for now.